src/celldega/pre/merge_segmentations.py

```diff
@@ -0,0 +1,514 @@
+
+    """
+    Adds a polygon to the no-conflict GeoDataFrame or merges if overlap exceeds threshold.
+
+    Args:
+        gdf_nc (GeoDataFrame): GeoDataFrame of non-conflicting polygons.
+        default_clustering (DataFrame): Default clustering information.
+        clusters_within_cutout_region (list): Clusters within the region of interest.
+        poly (Polygon): Polygon to be added or merged.
+        ioa_thresh (float): Threshold for intersection over area.
+        ID (str): Cell ID for the new polygon.
+        tech (str): Technology or source name.
+
+    Returns:
+        GeoDataFrame: Updated gdf_nc with new or merged polygon.
+    """
+
+    possible_intersections = gdf_nc.sindex.query(poly, predicate='intersects')
+
+    if len(possible_intersections) == 0:
+
+        new_data = {
+            'geometry': poly
+        }
+
+        new_row = gpd.GeoDataFrame([new_data])
+        new_row['cell_id'] = ID
+        new_row['technology'] = tech
+        new_row['centroid'] = poly.centroid
+        gdf_nc = gpd.GeoDataFrame(pd.concat([gdf_nc, new_row], ignore_index=True))
+
+    else:
+
+        max_ioa_merged = 0
+        max_ioa_merged_index = 0
+
+        if not poly.is_valid:
+            poly = make_valid(poly)
+
+        for index in possible_intersections:
+
+            poly_intersect = gdf_nc.iloc[index]['geometry']
+
+            if not poly_intersect.is_valid:
+                poly_intersect = make_valid(poly_intersect).buffer(0)
+                poly_intersect = poly_intersect.simplify(0.001)
+
+            if min(poly.area, poly_intersect.area) > 0:
+                ioa_merged = poly_intersect.intersection(poly).area / min(poly.area, poly_intersect.area)
+            else:
+                ioa_merged = 0
+
+            if ioa_merged >= max_ioa_merged:
+                max_ioa_merged = ioa_merged
+                max_ioa_merged_index = index
+
+        if max_ioa_merged >= ioa_thresh:
+
+            poly_intersect = gdf_nc.loc[max_ioa_merged_index, 'geometry']
+
+            poly_intersect = make_valid(poly_intersect).buffer(0)
+            poly_intersect = poly_intersect.simplify(0.001)
+
+            if gdf_nc.loc[max_ioa_merged_index, 'cell_id'] in default_clustering[~default_clustering['Cluster'].isin(clusters_within_cutout_region)].index.to_list():
+                return gdf_nc
+
+            else:
+                gdf_nc = gdf_nc.drop(max_ioa_merged_index)
+
+        else:
+
+            new_data = {'geometry': poly}
+            new_row = gpd.GeoDataFrame([new_data])
+            new_row['cell_id'] = ID
+            new_row['technology'] = tech
+            new_row['centroid'] = poly.centroid
+            gdf_nc = gpd.GeoDataFrame(pd.concat([gdf_nc, new_row], ignore_index=True))
+
+    return gdf_nc
+
+def merge_segmentation(default_data_path, custom_data_path, output_path, clusters_within_cutout_region, inv_alpha_value_for_cutout_region=1, buffer_for_cutout_region_alpha_shape=0, ioa_small_thresh = 0.5):
```

Comment:
Consider adding some comments to show what is going on, especially between L283-L392. Most of the lines with `intersecting_cell_pairs_before_harmonization_DF` are pretty long, almost too long. No big deal but maybe you can try `black` it and see if it works?